tf/tfprocess.py

```diff
@@ -1135,61 +1197,122 @@ def residual_block(self, inputs, channels, name):
+        # (batch_size, num_heads, 64, depth)
+        return tf.transpose(reshaped, perm=[0, 2, 1, 3])
+
+    def scaled_dot_product_attention(self, q, k, v, name: str = None, inputs=None):
 
-    def scaled_dot_product_attention(self, q, k, v):
+        # 0 h 64 d, 0 h d 64
         matmul_qk = tf.matmul(q, k, transpose_b=True)
         dk = tf.cast(tf.shape(k)[-1], self.model_dtype)
         scaled_attention_logits = matmul_qk / tf.math.sqrt(dk)
+        heads = scaled_attention_logits.shape[1]
+
+        if self.use_smolgen:
+            smolgen_weights = self.smolgen_weights(inputs, heads, self.smolgen_hidden_channels, self.smolgen_hidden_sz,
+                                                   self.smolgen_gen_sz, name=name+'/smolgen', activation=self.smolgen_activation)
+            scaled_attention_logits = scaled_attention_logits + smolgen_weights
+
         attention_weights = tf.nn.softmax(scaled_attention_logits, axis=-1)
         output = tf.matmul(attention_weights, v)
-        return output, scaled_attention_logits
 
+        # output shape = (b, h, 64, d)
+        if heads > 1:
+            output = tf.transpose(output, perm=[0, 2, 1, 3])
+            output = tf.keras.layers.Reshape((64, -1))(output)
+        return output, scaled_attention_logits
     # multi-head attention in encoder blocks
-    def mha(self, inputs, emb_size, d_model, num_heads, name):
+    
+    def mha(self, inputs, emb_size: int, d_model: int, num_heads: int, initializer, name: str):
         assert d_model % num_heads == 0
         depth = d_model // num_heads
         # query, key, and value vectors for self-attention
-        q = tf.keras.layers.Dense(d_model, kernel_initializer='glorot_normal', name=name + '/wq')(inputs)
-        k = tf.keras.layers.Dense(d_model, kernel_initializer='glorot_normal', name=name + '/wk')(inputs)
-        v = tf.keras.layers.Dense(d_model, kernel_initializer='glorot_normal', name=name + '/wv')(inputs)
+        # inputs b, 64, sz
+
+        q = tf.keras.layers.Dense(
+            d_model, name=name+'/wq', kernel_initializer='glorot_normal')(inputs)
+        k = tf.keras.layers.Dense(
+            d_model, name=name+'/wk', kernel_initializer='glorot_normal')(inputs)
+        v = tf.keras.layers.Dense(
+            d_model, name=name+'/wv', kernel_initializer=initializer)(inputs)
+
         # split q, k and v into smaller vectors of size 'depth' -- one for each head in multi-head attention
         batch_size = tf.shape(q)[0]
+
         q = self.split_heads(q, batch_size, num_heads, depth)
         k = self.split_heads(k, batch_size, num_heads, depth)
         v = self.split_heads(v, batch_size, num_heads, depth)
-        scaled_attention, attention_weights = self.scaled_dot_product_attention(q, k, v)
-        if num_heads > 1:
-            scaled_attention = tf.transpose(scaled_attention, perm=[0, 2, 1, 3])
-            scaled_attention = tf.reshape(scaled_attention, (batch_size, -1, d_model))  # concatenate heads
```

Comment:
Curious this reshape is not the same as the one now embedded into the called function...

Reply:
reverted back, not sure if there is a difference in the end